security/pkg/platform/aws_test.go

```diff
@@ -56,39 +61,286 @@ func TestIsProperPlatform(t *testing.T) {
 	}
 }
 
-func TestGetInstanceIdentityDocument(t *testing.T) {
-	sigFile := "testdata/sig.pem"
-	sigBytes, err := ioutil.ReadFile(sigFile)
-	if err != nil {
-		t.Fatalf("unable to read file %s", sigFile)
+func TestNewAwsClientImpl(t *testing.T) {
+	client := NewAwsClientImpl(AwsConfig{})
+	if client == nil {
+		t.Errorf("NewAwsClientImpl should not return nil")
 	}
+}
 
-	server := initTestServer(
-		"/latest/dynamic/instance-identity/pkcs7",
-		sigBytes,
-	)
-	defer server.Close()
+func TestAwsGetInstanceIdentityDocument(t *testing.T) {
+	testCases := map[string]struct {
+		sigFile              string
+		requestPath          string
+		expectedErr          string
+		expectedInstanceType string
+		expectedRegion       string
+		expectedCredential   string
+	}{
+		"Good Identity": {
+			sigFile:              "testdata/sig.pem",
+			requestPath:          "/latest/dynamic/instance-identity/pkcs7",
+			expectedErr:          "",
+			expectedInstanceType: "c3.large",
+			expectedRegion:       "us-west-2",
+			expectedCredential: "\"ewogICJkZXZwYXlQcm9kdWN0Q29kZXMiIDogbnVsbCwKICAicHJpdmF0ZUlwIiA6ICIx" +
+				"MC4xNi4xNy4yNDgiLAogICJhdmFpbGFiaWxpdHlab25lIiA6ICJ1cy13ZXN0LTJiIiwKICAidmVyc2lvbiIgOi" +
+				"AiMjAxMC0wOC0zMSIsCiAgImluc3RhbmNlSWQiIDogImktMDY0NmM5ZWZlMmU2MmRjNjMiLAogICJiaWxsaW5n" +
+				"UHJvZHVjdHMiIDogbnVsbCwKICAiaW5zdGFuY2VUeXBlIiA6ICJjMy5sYXJnZSIsCiAgImFjY291bnRJZCIgOi" +
+				"AiOTc3Nzc3NjU3NjExIiwKICAiYXJjaGl0ZWN0dXJlIiA6ICJ4ODZfNjQiLAogICJrZXJuZWxJZCIgOiBudWxs" +
+				"LAogICJyYW1kaXNrSWQiIDogbnVsbCwKICAiaW1hZ2VJZCIgOiAiYW1pLWZhYmY1YzgyIiwKICAicGVuZGluZ1" +
+				"RpbWUiIDogIjIwMTctMDgtMjdUMTc6MTg6MjBaIiwKICAicmVnaW9uIiA6ICJ1cy13ZXN0LTIiCn0=\"",
+		},
+	}
 
-	c := ec2metadata.New(unit.Session, &aws.Config{Endpoint: aws.String(server.URL + "/latest")})
-	na := &AwsClientImpl{client: c}
+	for id, c := range testCases {
+		sigBytes, err := ioutil.ReadFile(c.sigFile)
+		assert.Equal(t, err, nil, fmt.Sprintf("%v: Unable to read file %s", id, c.sigFile))
+
+		server := initTestServer(
+			c.requestPath,
+			sigBytes,
+		)
+		defer server.Close()
+
+		awsc := &AwsClientImpl{
+			client: ec2metadata.New(unit.Session, &aws.Config{Endpoint: aws.String(server.URL + "/latest")}),
+		}
+
+		docBytes, err := awsc.getInstanceIdentityDocument()
+		if len(c.expectedErr) > 0 {
```

Comment:
In general I would try to avoid complicated logic in test. How about having 2 test function, one just for testing for error, and other to test the output?

Reply:
This is also a very common pattern in your code. May be have a compareError helper function then the code would be easier to follow.